Make Custom Metrics Basic Authentication Optional (#5)
SAP-docs/btp-application-autoscaler: docs/defining-a-custom-metric-87e657e.md

As an alternative to the standard metrics provided by the Application Autoscaler, you can also define custom metrics such as memory consumed in megabytes or in percentage, response time, and throughput for scaling. Custom metrics offer more flexibility, so that you can scale applications based on your own requirements.

> To use custom metrics, your application need to perform the following tasks:

Reviewer: ```suggestion
To use custom metrics, your application needs to perform the following tasks:
```

     ```
     # sample policy.json with credential Type as x509
    {
        "instance_min_count":1,
        "instance_max_count":4,
        "scaling_rules":[
            {
            "metric_type":"jobqueue",
            "breach_duration_secs":60,
            "threshold":100,
            "operator":">=",
            "cool_down_secs":120,
            "adjustment":"+1"
            }
        ],
        "credential-type": "x509"
    }
    
    # bind the service instance with the scaling policy
    cf bind-service <application> <application-autoscaler-service-instance> -c policy.json
    ```

2.  Read custom metrics binding credentials from the application environment.

    > ###  Binding Credentials for mTLS
    > ```
    > "custom_metrics": {
    >       "mtls_url": "https://autoscaler-metrics-mtls.cf.<landscape>.hana.ondemand.com",
    >      }
    > ```

3.  **Authenticate with X.509 Certificate**: Your application requires base url, certificate and a private key for mutual TLS authentication \(mTLS\).

    - use the base URL provided in the `mtls_url` of the service binding

>     - use the X.509 certificate and private key from the paths in `CF_INSTANCE_CERT` and `CF_INSTANCE_KEY` respectively.

Reviewer: ```suggestion
    - use the X.509 certificate and private key from the paths provided in the environment variables `CF_INSTANCE_CERT` and `CF_INSTANCE_KEY` respectively.
```

>     >       ### Note:

Reviewer: Maybe we can also answer a question I've had to answer a few times:

```suggestion
    >       ### Notes:
             The environment variables `CF_INSTANCE_CERT` and `CF_INSTANCE_KEY` are not shown in the cockpit or using the `cf env` command. They are only visible inside the running application container.
```

    >       The X.509 certificate and private key pair are valid for 24 hours. At least 20 minutes before expiration, they are regenerated and new files replace the existing files. Make sure that your code reloads the X.509 certificate and private key pair if they are expired. See [Using Instance Identity Credentials](https://docs.cloudfoundry.org/devguide/deploy-apps/instance-identity.html) in the Cloud Foundry Documentation.

>     **Authenticate with Basic Authentication (Not Recommended)**: `url`, `username`, and `password` in the application environment are provided for backwards compatibility with the HTTP basic authentication scheme and should not be used going forward. If your application still need to authenticate with basic authentication, use credential-type 'binding-secret' in the service binding.

Reviewer: ```suggestion
    **Authenticate with Basic Authentication (Deprecated, do not use)**: `url`, `username`, and `password` in the application environment are provided for backwards compatibility with the HTTP basic authentication scheme and should not be used going forward. If your application still needs to authenticate with basic authentication, use credential-type `binding-secret` in the service binding and plan to migrate to mTLS (`x509`) authentication until 2025-01-23.
```
